modules/core/src/test/java/org/apache/ignite/internal/processors/cache/datastructures/IgniteReentrantLockTest.java

```diff
@@ -0,0 +1,318 @@
+     * @throws Exception If failed.
+     */
+    public void testReentrantLock() throws Exception {
+        testReentrantLock(true);
+        testReentrantLock(false);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testReentrantLock(final boolean fair) throws Exception {
+        for (int i = 0; i < NODES_CNT; i++) {
+            IgniteLock lock = createReentrantLock(i, fair ? "lock1" : "lock2", fair);
+
+            // Other node can still don't see update.
+            GridTestUtils.waitForCondition(new GridAbsPredicate() {
+                @Override public boolean apply() {
+                    return !lock.isLocked();
+                }
+            }, 500L);
+
+            lock.lock();
+
+            assertTrue(lock.isLocked());
+
+            lock.lock();
+
+            assertTrue(lock.isLocked());
+
+            lock.unlock();
+
+            assertTrue(lock.isLocked());
+
+            lock.unlock();
+
+            assertFalse(lock.isLocked());
+        }
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinode() throws Exception {
+        testReentrantLockMultinode(false);
+        testReentrantLockMultinode(true);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinode(final boolean fair) throws Exception {
+        List<IgniteInternalFuture<Void>> futs = new ArrayList<>();
+
+        final AtomicInteger inx = new AtomicInteger(0);
+
+        GridTestUtils.runMultiThreadedAsync(new Runnable() {
+            @Override public void run() {
+                try {
+                    IgniteLock lock = createReentrantLock(inx.getAndIncrement(), fair ? "lock1" : "lock2", fair);
+                    lock.lock();
+
+                    try {
+                        assertTrue(lock.isLocked());
+                        Thread.sleep(1_000L);
+                    }
+                    catch (InterruptedException e) {
+                        e.printStackTrace();
+                    }
+                    finally {
+                        lock.unlock();
+                    }
+                }
+                catch (Exception e) {
+                    e.printStackTrace();
+                }
+            }
+        }, NODES_CNT, "worker").get(30_000L);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinodeFailoverUnfair() throws Exception {
+        testReentrantLockMultinodeFailover(false);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinodeFailoverFair() throws Exception {
+        testReentrantLockMultinodeFailover(true);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinodeFailover(final boolean fair) throws Exception {
+        List<IgniteInternalFuture<Void>> futs = new ArrayList<>(NODES_CNT);
+
+        for (int i = 0; i < NODES_CNT; i++) {
```

Comment:
may be use GridTestUtils.runMultiThreaded()

Reply:
fixed